pkg/tidb/client.go

```diff
@@ -81,40 +83,50 @@ func NewTiDBClient(lc fx.Lifecycle, config *config.Config, etcdClient *clientv3.
 	return client
 }
 
-func (c *Client) WithStatusAPITimeout(timeout time.Duration) *Client {
-	c2 := *c
-	c2.statusAPITimeout = timeout
-	return &c2
+func (c Client) WithStatusAPITimeout(timeout time.Duration) *Client {
+	c.statusAPITimeout = timeout
+	return &c
 }
 
-func (c *Client) WithStatusAPIAddress(host string, statusPort int) *Client {
-	c2 := *c
-	c2.statusAPIAddress = fmt.Sprintf("%s:%d", host, statusPort)
-	return &c2
+func (c Client) WithStatusAPIAddress(host string, statusPort int) *Client {
+	c.statusAPIAddress = fmt.Sprintf("%s:%d", host, statusPort)
+	return &c
 }
 
-func (c *Client) WithSQLAPIAddress(host string, sqlPort int) *Client {
-	c2 := *c
-	c2.sqlAPIAddress = fmt.Sprintf("%s:%d", host, sqlPort)
-	return &c2
+func (c Client) WithEnforcedStatusAPIAddress(host string, statusPort int) *Client {
+	c.statusAPIAddress = fmt.Sprintf("%s:%d", host, statusPort)
+	c.enforcedSettings.Add(EnforcedSettingStatusAPIAddress)
+	return &c
+}
+
+func (c Client) WithSQLAPIAddress(host string, sqlPort int) *Client {
+	c.sqlAPIAddress = fmt.Sprintf("%s:%d", host, sqlPort)
+	return &c
 }
 
 func (c *Client) OpenSQLConn(user string, pass string) (*gorm.DB, error) {
+	var err error
+
 	overrideEndpoint := os.Getenv(tidbOverrideSQLEndpointEnvVar)
-	if overrideEndpoint != "" && c.sqlAPIAddress != "" {
+	hasEnforcedSettingSQLAPIAddress := c.enforcedSettings.Has(EnforcedSettingSQLAPIAddress)
+	// the `tidbOverrideSQLEndpointEnvVar` and the `Client.sqlAPIAddress` have the same override priority, if both exist and have not enforced `Client.sqlAPIAddress` then an error is returned
+	if overrideEndpoint != "" && c.sqlAPIAddress != "" && !hasEnforcedSettingSQLAPIAddress {
 		log.Warn(fmt.Sprintf("Reject to establish a target specified TiDB SQL connection since `%s` is set", tidbOverrideSQLEndpointEnvVar))
 		return nil, ErrTiDBConnFailed.New("TiDB Dashboard is configured to only connect to specified TiDB host")
 	}
```

Comment:
How about moving the check logic into `WithSQLAPIAddress`? Then you don't need to store the enforcedSettings any more.

Reply:
Does this mean I need to store errors for chained api?